src/commands/version.js

```diff
@@ -0,0 +1,275 @@
+const dirsExist = require('../utils/dirs-exist');
+const { terminal } = require('terminal-kit');
+const fs = require('fs');
+const path = require('path'); // Path module
+const semver = require('semver');
+
+exports.command = 'version [type] [projects]';
+exports.desc = 'Run an npm version bump process.';
+exports.builder = (yargs) => {
+  yargs.option('type', {
+    describe: 'type of version to bump.',
+    default: 'string',
+    type: 'string',
+  });
+
+  yargs.positional('projects', {
+    describe: 'Comma separated list of projects to compile.',
+    type: 'string',
+    default: '',
+  });
+
+  yargs.option('site', {
+    describe: `Run the process from the root of a site, such as from wp-content.`,
+    default: false,
+    type: 'boolean',
+  });
+};
+exports.handler = async ({ type = '', projects = '', site = false }) => {
+  // with no projects listed it returns a empty array
+  const projectsList = projects.split(',').filter((item) => item.length > 0);
+  const excludeDirs = ['node_modules', 'dist', 'build', 'vendor', 'vendor_prefixed'];
+  const targetFiles = ['package.json', 'style.css', 'plugin.php'];
+
+  const hasTargetDirs = dirsExist(targetDirs);
+
+  const isAllProjects = (site || hasTargetDirs) && !projects;
+
+  /**
+   * Function to search for all files which are not in the excluded directories
+   *
+   * @param {string} basePath - The base directory to start searching from.
+   * @param {Array} directoryNames - An array of directory names to match.
+   * @returns {Array} - An array of matching directory paths.
+   */
+  const getAllFiles = (dirPath, arrayOfFiles) => {
+    const files = fs.readdirSync(dirPath);
+
+    arrayOfFiles = arrayOfFiles || [];
+
+    files.forEach((file) => {
+      const filePath = path.join(dirPath, file);
+      if (fs.statSync(filePath).isDirectory()) {
+        if (!excludeDirs.includes(file)) {
+          arrayOfFiles = getAllFiles(filePath, arrayOfFiles);
+        }
+      } else {
+        arrayOfFiles.push(filePath);
+      }
+    });
+
+    return arrayOfFiles;
+  };
+
+  /**
+   * Function to find the files we want to use.
+   * @param {string} baseDirs - Base directory to search from.
+   * @param {Array} targets - Files we want to find
+   */
+  const findTargetFiles = (baseDirs, targets) => {
+    baseDirs.forEach((baseDir) => {
+      if (fs.existsSync(baseDir) && fs.statSync(baseDir).isDirectory()) {
+        const allFiles = getAllFiles(baseDir);
+        allFiles.forEach((filePath) => {
+          const fileName = path.basename(filePath);
+          if (targets.includes(fileName)) {
+            incrementVersionNumber(fileName, filePath, type);
+          }
+        });
+      } else {
+        console.error(`Directory not found: ${baseDir}`);
+      }
+    });
+  };
```

Comment:
We don't need to recursively seek out files in this manner. This will add a lot of overhead when creating compiling the list, especially in larger projects where theres multiple layers and directories. You can reduce this down into something more performant by using the project directories and targeting those, as we already know that all the files required will live in the project directory.

For example these are what you need to be looking for at a minimum:

| File                                |
| ----------------------------------- |
| `example-plugin/*.php`         |
| `example-plugin/style.css` |
| `example-plugin/package.json`      |
| `example-plugin/package-lock.json`      |

Reply:
@ampersarnie would it be possible to get another example on this please im struggling to get my head around this.   My thoughts are that you would need to recursively search to find the plugins ?

Reply:
If you look at the [`build` command file](https://github.com/bigbite/build-tools/blob/bc6488a2fe4403e659ad2ad6b2071ab898598808/src/commands/build.js#L62) you will see that it already holds functionality for finding plugins/themes that Build Tools can and should process. Lines 62 through to 113 handle all of the looking up projects and creating that list for you to work from. It provides a whole range of paths and file points that you can use, such as the `package.json`.

Because we know that the files you need to increment are in the root of a project (i.e. `plugins/my-plugin/*`), you can target files within that directory alone, removing the need for recursion.

As for why not use recursion, there are a number of them:
1. It can add unnecessary processing to the file look up as its not just targeting files, or directory levels that we are aware of (the root level). But it's also going to look through every other directory in the plugin. For example I was able to get a file processed that I added to `dir1/dir2/dir3/some-file.php`. It doesn't need it for the case we have here.
2. In conjunction with [the changes you made here](https://github.com/bigbite/build-tools/pull/119#discussion_r1656929382) and after further testing, it is also possible due to the recursion setup to have a plugin and/or theme that **should not** be processed by Build Tools. This is because the recursion and lookup is targeting any file and any directory that has a `package.json` file. This is going to cause issues with updating versions in 3rd party plugins/themes. Take the following steps:
- Create a plugin package.json - `plugins/non-build-tools-plugin/package.json` - and add the following basic config:
```json
{
  "version": "12.3.4"
}
```
- Create a plugin entry file - `plugins/non-build-tools-plugin/plugin.php` - and add the following:
```php
<?php
/**
 * Version: 12.3.4
 **/
```
- Run a version update on all projects from `wp-content`
```bash
build-tools version patch
```
- You should see `non-build-tools-plugin` being processed in the list.

That dummy project you have created **should not** have a version increase as it doesn't have the correct setup to work across our build setup. You will also see the same results if you add a third-party plugin/theme to one of the related directories - `client-mu-plugins`, `plugins` or `themes`.

Does this help?